pyomo/contrib/mindtpy/single_tree.py

```diff
@@ -0,0 +1,257 @@
+                        opt._pyomo_var_to_solver_var_map[var])) for var in constr_vars) + value(constr.body)
+                    cplex_rhs = -generate_standard_repn(pyomo_expr).constant
+                    cplex_expr, _ = opt._get_expr_from_pyomo_expr(pyomo_expr)
+                    self.add(constraint=cplex.SparsePair(ind=cplex_expr.variables, val=cplex_expr.coefficients),
+                             sense="G",
+                             rhs=constr.lower.value + cplex_rhs)
+
+    def handle_lazy_master_mip_feasible_sol(self, master_mip, solve_data, config, opt):
+        """ This function is called during the branch and bound of master mip, more exactly when a feasible solution is found and LazyCallback is activated.
+        Copy the result to working model and update upper or lower bound
+        In LP-NLP, upper or lower bound are updated during solving the master problem
+        """
+        # proceed. Just need integer values
+        MindtPy = master_mip.MindtPy_utils
+        main_objective = next(
+            master_mip.component_data_objects(Objective, active=True))
+
+        # this value copy is useful since we need to fix subproblem based on the solution of the master problem
+        self.copy_lazy_var_list_values(opt,
+                                       master_mip.MindtPy_utils.variable_list,
+                                       solve_data.working_model.MindtPy_utils.variable_list,
+                                       config)
+        # update the bound
+        if main_objective.sense == minimize:
+            solve_data.LB = max(
+                self.get_objective_value(),
+                # self.get_best_objective_value(),
+                solve_data.LB)
+            solve_data.LB_progress.append(solve_data.LB)
+        else:
+            solve_data.UB = min(
+                self.get_objective_value(),
+                # self.get_best_objective_value(),
+                solve_data.UB)
+            solve_data.UB_progress.append(solve_data.UB)
+        config.logger.info(
+            'MIP %s: OBJ: %s  LB: %s  UB: %s'
+            % (solve_data.mip_iter, value(MindtPy.MindtPy_oa_obj.expr),
+               solve_data.LB, solve_data.UB))
+
+    def handle_lazy_NLP_subproblem_optimal(self, fix_nlp, solve_data, config, opt):
+        """Copies result to mip(explaination see below), updates bound, adds OA and integer cut,
+        stores best solution if new one is best"""
+        for c in fix_nlp.tmp_duals:
+            if fix_nlp.dual.get(c, None) is None:
+                fix_nlp.dual[c] = fix_nlp.tmp_duals[c]
+        dual_values = list(fix_nlp.dual[c]
+                           for c in fix_nlp.MindtPy_utils.constraint_list)
+
+        main_objective = next(
+            fix_nlp.component_data_objects(Objective, active=True))
+        if main_objective.sense == minimize:
+            solve_data.UB = min(value(main_objective.expr), solve_data.UB)
```

Comment:
In line 149 you use `self.get_objective_value()` - can we do the same here?

Reply:
`self.get_objective_value()` is an function from cplex package. It is used to get the objective value of the current integer solution in the branch and bound tree. When solving master problem, we need to use `self.get_objective_value()`. However, when we are solving subproblems, we still need to use the functions in pyomo.